lib/auth/userloginstate/generator.go

```diff
@@ -364,3 +431,49 @@ func (g *Generator) LoginHook(ulsService services.UserLoginStates) func(context.
 		return trace.Wrap(err)
 	}
 }
+
+// validateRoles is a helper function which validates that a given list of roles exist.
+func (g *Generator) validateRoles(ctx context.Context, roles []string) ([]string, []error, error) {
```

Comment:
Do we really need to return the `notFoundErrs` we basically know they're going to be "role xyz not found" right?

Also I would explore different names here. I'd expect something named validate to just return an error. Maybe this should be named something like `identifyMissingRoles`?